pytorch_analysis.ipynb

```diff
@@ -393,9 +404,318 @@
+    "# extract coefficients by method/seed/fold\n",
+    "for signal in ['signal', 'shuffled']:\n",
+    "    torch_coefs_folds[signal] = {}\n",
+    "    sk_coefs_folds[signal] = {}\n",
+    "    for seed in seeds:\n",
+    "        torch_coefs_folds[signal][seed] = {}\n",
+    "        sk_coefs_folds[signal][seed] = {}\n",
+    "        sk_coefs_file = os.path.join(results_dir,\n",
+    "                                     'sklearn_coefs_{}_{}.tsv.gz'.format(signal, seed))\n",
+    "        sk_coefs_df = pd.read_csv(sk_coefs_file, sep='\\t').sort_values('feature')\n",
+    "        torch_coefs_file = os.path.join(results_dir,\n",
+    "                                        'torch_coefs_{}_{}.tsv.gz'.format(signal, seed))\n",
+    "        torch_coefs_df = pd.read_csv(torch_coefs_file, sep='\\t').sort_values('feature')\n",
+    "        folds = np.unique(sk_coefs_df['fold'].values)\n",
+    "        for fold in folds:\n",
+    "            sk_coefs = sk_coefs_df[sk_coefs_df['fold'] == fold]['weight'].values\n",
+    "            sk_genes = sk_coefs_df[sk_coefs_df['fold'] == fold]['feature'].values\n",
+    "            torch_coefs = torch_coefs_df[torch_coefs_df['fold'] == fold]['weight'].values\n",
+    "            torch_genes = torch_coefs_df[torch_coefs_df['fold'] == fold]['feature'].values\n",
+    "            # for pytorch, to avoid dealing with numerical error near 0, just pick a\n",
+    "            # (somewhat arbitrary) threshold and set all smaller coefficients to 0\n",
+    "            close_to_zero = (np.abs(torch_coefs) < (np.abs(torch_coefs).max() / 1000))\n",
+    "            torch_coefs[close_to_zero] = 0\n",
+    "            torch_coefs_folds[signal][seed][fold] = (torch_coefs, torch_genes)\n",
+    "            sk_coefs_folds[signal][seed][fold] = (sk_coefs, sk_genes)\n",
+    "        \n",
+    "# jaccard similarity (works with unequal length strings)\n",
+    "def compute_jaccard(v1, v2):\n",
+    "    intersection = v1.intersection(v2)\n",
+    "    union = v1.union(v2)\n",
+    "    return (len(intersection) / len(union) if len(union) != 0 else 0)\n",
+    "        \n",
+    "# get inter-method similarity on same seed/fold\n",
+    "inter_method_sims = []\n",
+    "for seed in seeds:\n",
+    "    for fold in folds:\n",
+    "        sk_coefs = sk_coefs_folds['signal'][seed][fold][0]\n",
+    "        sk_genes = sk_coefs_folds['signal'][seed][fold][1]\n",
+    "        sk_nz_coefs = (sk_coefs != 0)\n",
+    "        sk_nz_genes = sk_genes[sk_nz_coefs]\n",
+    "        torch_coefs = torch_coefs_folds['signal'][seed][fold][0]\n",
+    "        torch_genes = torch_coefs_folds['signal'][seed][fold][1]\n",
+    "        torch_nz_coefs = (torch_coefs != 0)\n",
+    "        torch_nz_genes = torch_genes[torch_nz_coefs]\n",
+    "        inter_method_sims.append(compute_jaccard(set(sk_nz_genes), set(torch_nz_genes)))\n",
+    "        \n",
+    "import itertools as it\n",
```

Comment:
This block of code seems very similar (but not identical) to the same block 50 lines down. Could they be combined into a single loop? If nothing else, you shouldn't have to import itertools twice.

Reply:
I moved this logic into a utilities file and got rid of the duplication - let me know if that's what you had in mind.